compiler/src/dotty/tools/dotc/transform/Constructors.scala

```diff
@@ -149,17 +149,21 @@ class Constructors extends MiniPhase with IdentityDenotTransformer { thisPhase =
     //  (2) If the parameter accessor reference was to an alias getter,
     //      drop the () when replacing by the parameter.
     object intoConstr extends TreeMap {
-      private var isSuperCall = false
+      private var inSuperCall = false
       override def transform(tree: Tree)(using Context): Tree = tree match {
         case Ident(_) | Select(This(_), _) =>
           var sym = tree.symbol
-          if sym.is(ParamAccessor) && (!sym.is(Mutable) || isSuperCall)
-            // Variables need to go through the getter since they might have been updated,
-            // except if we are in a super call, since then the virtual getter call would
-            // be illegal.
-          then
+          def isOverridableSelect = tree.isInstanceOf[Select] && !sym.isEffectivelyFinal
+          def keepUnlessInSuperCall = sym.is(Mutable) || isOverridableSelect
+            // If true, switch to constructor parameters only in the super call.
+            // Variables need to go through the getter since they might have been updated.
+            // References via this need to use the getter as well as long as that getter
+            // can be overriddem. This is needed to handle overrides correctly. See run/i15723.scala.
```

Comment:
typo overridden.

I understand the nice comment but not `keepUnlessInSupercall`, which means "keep accessor but you must also check inSuperCall". The double negative `!unless` is tricky. `def requiresParamSyms = inSuperCall || !sym.is(Mutable) && !isOverridableSelect` which just distributes the negation.

This PR still permits `class C(x: Int) { def f = this.x }` use of selection for parameter.

Reply:
Good point. I turned the sense around.